src/kernel/include/pgtbl.h

```diff
@@ -501,53 +698,73 @@ extern unsigned long __cr3_contents;
 pgtbl_update(pgtbl_t pt)
-{ asm volatile("mov %0, %%cr3" : : "r"(pt)); }
+{
+	/* PRY:This page table must be the top-level page table that contains the MPU data */
+	/* Do MPU data loading here */
+	/*asm volatile("mov %0, %%cr3" : : "r"(pt));*/
+}
 
 /* vaddr -> kaddr */
 static vaddr_t
 pgtbl_translate(pgtbl_t pt, u32_t addr, u32_t *flags)
-{ return (vaddr_t)pgtbl_lkup(pt, addr, flags); }
+{
+	u32_t* pte;
+	u32_t size_order;
+	pte=_pgtbl_lookup(pt, addr, &size_order);
+	if(pte==0)
+		return 0;
+
+	/* If the pte is there, we return it */
+	if(*pte!=0)
+		return COS_PGTBL_METAADDR(*pte);
+	else
+		return 0;
+}
 
 /* FIXME: this should be using cos_config.h defines */
 #define KERNEL_PGD_REGION_OFFSET  (PAGE_SIZE - PAGE_SIZE/4)
 #define KERNEL_PGD_REGION_SIZE    (PAGE_SIZE/4)
 
-static pgtbl_t pgtbl_create(void *page, void *curr_pgtbl) {
-	pgtbl_t ret = pgtbl_alloc(page);
-	/* Copying the kernel part of the pgd. */
-	memcpy(page + KERNEL_PGD_REGION_OFFSET, (void *)chal_pa2va((paddr_t)curr_pgtbl) + KERNEL_PGD_REGION_OFFSET, KERNEL_PGD_REGION_SIZE);
-
-	return ret;
-}
-int pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, pgtbl_t pgtbl, u32_t lvl);
+/* PRY:this function no longer called because all nodes are embedded */
+//static pgtbl_t pgtbl_create(void *page, void *curr_pgtbl) {
+//	pgtbl_t ret = pgtbl_alloc(page);
+//	/* Copying the kernel part of the pgd. */
+//	memcpy(page + KERNEL_PGD_REGION_OFFSET, (void *)chal_pa2va((paddr_t)curr_pgtbl) + KERNEL_PGD_REGION_OFFSET, KERNEL_PGD_REGION_SIZE);
+//
+//	return ret;
+//}
+/* PRY: modified the function definitions here */
+int //__attribute__((optimize("O0")))
+pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, u32_t type, u32_t start_addr, u32_t size_order, u32_t num_order);
+/* int pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, pgtbl_t pgtbl, u32_t lvl); */
 int pgtbl_deactivate(struct captbl *t, struct cap_captbl *dest_ct_cap, unsigned long capin,
 		     livenessid_t lid, capid_t pgtbl_cap, capid_t cosframe_addr, const int root);
 
 static int
 pgtbl_mapping_scan(struct cap_pgtbl *pt)
 {
-	unsigned int i, pte, *page;
-	livenessid_t lid;
-	u64_t past_ts;
-
-	/* This scans the leaf level of the pgtbl and verifies
-	 * quiescence. */
-	if (pt->lvl != PGTBL_DEPTH - 1) return -EINVAL;
-
-	page = (unsigned int *)(pt->pgtbl);
-	assert(page);
-
-	for (i = 0; i < PAGE_SIZE / sizeof(int *); i++) {
-		pte = *(page + i);
-		if (pte & PGTBL_PRESENT || pte & PGTBL_COSFRAME) return -EINVAL;
-
-		if (pte & PGTBL_QUIESCENCE) {
-			lid = pte >> PGTBL_PAGEIDX_SHIFT;
-
-			if (ltbl_get_timestamp(lid, &past_ts)) return -EFAULT;
-			if (!tlb_quiescence_check(past_ts)) return -EQUIESCENCE;
-		}
-	}
+//	unsigned int i, pte, *page;
+//	livenessid_t lid;
+//	u64_t past_ts;
+//
+//	/* This scans the leaf level of the pgtbl and verifies
+//	 * quiescence. - we are always quiescent, so return 0 */
+//	if (pt->lvl != PGTBL_DEPTH - 1) return -EINVAL;
+//
```

Comment:
not clear if removing all of this is going to break something.  We need to know.